core/src/autogluon/core/hpo/resources_calculator.py

```diff
@@ -0,0 +1,172 @@
+        minimum_cpu_per_trial,
+        model_estimate_memory_usage,
+        **kwargs,
+    ):
+        cpu_per_job = max(minimum_cpu_per_trial, int(total_num_cpus // num_samples))
+        max_jobs_in_parallel_memory = num_samples
+
+        if model_estimate_memory_usage is not None:
+            mem_available = psutil.virtual_memory().available
+            # calculate how many jobs can run in parallel given memory available
+            max_jobs_in_parallel_memory = max(1, int(mem_available // model_estimate_memory_usage))
+        num_parallel_jobs = min(num_samples, total_num_cpus // cpu_per_job, max_jobs_in_parallel_memory)
+        cpu_per_job = max(minimum_cpu_per_trial, int(total_num_cpus // num_parallel_jobs))  # update cpu_per_job in case memory is not enough and can use more cores for each job
+        resources_per_trial = dict(cpu=cpu_per_job)
+        
+        return dict(
+                    resources_per_trial=resources_per_trial,
+                    num_parallel_jobs=num_parallel_jobs,
+                    cpu_per_job=cpu_per_job
+                )
+    
+    
+class GpuResourceCalculator(ResourceCalculator):
+    
+    @property
+    def type(self):
+        return 'gpu'
+    
+    def get_resources_per_trial(
+        self,
+        total_num_cpus,
+        total_num_gpus,
+        num_samples,
+        minimum_cpu_per_trial,
+        minimum_gpu_per_trial,
+        **kwargs,
+    ):
+        gpu_per_job = max(minimum_gpu_per_trial, total_num_gpus // num_samples)
+        num_parallel_jobs = total_num_gpus // gpu_per_job  # num_parallel_jobs purely based on gpu
+        cpu_per_job = max(minimum_cpu_per_trial, total_num_cpus // num_parallel_jobs)
+        num_parallel_jobs = min(num_parallel_jobs, total_num_cpus // cpu_per_job)  # update num_parallel_jobs in case cpu is not enough
+        gpu_per_job = max(minimum_gpu_per_trial, total_num_gpus // num_parallel_jobs)  # update gpu_per_job in case cpu was the bottleneck
+        resources_per_trial = dict(cpu=cpu_per_job, gpu=gpu_per_job)
+        
+        return dict(
+                    resources_per_trial=resources_per_trial,
+                    num_parallel_jobs=num_parallel_jobs,
+                    cpu_per_job=cpu_per_job,
+                    gpu_per_job=gpu_per_job,
+                )
+    
+    
+class RayLightningCpuResourceCalculator(ResourceCalculator):
+    
+    @property
+    def type(self):
+        return 'ray_lightning_cpu'
+    
+    def get_resources_per_trial(
+        self,
+        total_num_cpus,
+        num_samples,
+        minimum_cpu_per_trial,
+        **kwargs,
+    ):
+        # TODO: for cpu case, is it better to have more workers or more cpus per worker?
+        cpu_per_job = max(minimum_cpu_per_trial, total_num_cpus // num_samples)
+        num_parallel_jobs = min(num_samples, total_num_cpus // cpu_per_job)
+        num_workers = max(minimum_cpu_per_trial, cpu_per_job - 1)  # 1 cpu for master process
+        cpu_per_worker = 1
```

Comment:
ditto: no memory assessments?

Reply:
automm doesn't currently has memory estimation, but maybe it's still a good idea to have the logic here in case